Core/AppRuntime/Source/AppRuntimeV8.cpp

```diff
@@ -1,73 +1,89 @@
-            v8::HandleScope isolate_handle_scope{isolate};
-            v8::Local<v8::Context> context = v8::Context::New(isolate);
-            v8::Context::Scope context_scope{context};
-
-            Napi::Env env = Napi::Attach(context);
-            Run(env);
-            Napi::Detach(env);
-        }
-
-        // Destroy the isolate.
-        // todo : GetArrayBufferAllocator not available?
-        // delete isolate->GetArrayBufferAllocator();
-        isolate->Dispose();
-    }
+#include "AppRuntime.h"
+
+#ifndef __clang__
+#pragma warning(disable : 4100 4267 4127)
+#endif
+#include <v8.h>
+#include <libplatform/libplatform.h>
+
+namespace Babylon
+{
+    namespace
+    {
+        class Module final
+        {
+        public:
+            Module(const char* executablePath)
+            {
+                v8::V8::InitializeICUDefaultLocation(executablePath);
+                v8::V8::InitializeExternalStartupData(executablePath);
+                m_platform = v8::platform::NewDefaultPlatform();
+                v8::V8::InitializePlatform(m_platform.get());
+                v8::V8::Initialize();
+            }
+
+            ~Module()
+            {
+                v8::V8::Dispose();
+                v8::V8::ShutdownPlatform();
+            }
+
+            static Module& Instance()
+            {
+                return *s_module;
+            }
+
+            static void Initialize(const char* executablePath)
+            {
+                if (s_module == nullptr)
+                {
+                    s_module = std::make_unique<Module>(executablePath);
+                }
+            }
+
+            void AddPlatformToJavascript(Napi::Env env)
+            {
+                JsRuntime::NativeObject::GetFromJavaScript(env)
+                    .Set("_V8Platform", Napi::External<v8::Platform>::New(env, m_platform.get()));
+            }
+
+        private:
+            std::unique_ptr<v8::Platform> m_platform;
+
+            static std::unique_ptr<Module> s_module;
+        };
+
+        std::unique_ptr<Module> Module::s_module;
+
+        unsigned short s_inspectorPort = 0;
+    }
+
+    void AppRuntime::RunEnvironmentTier(const char* executablePath)
+    {
+        // Create the isolate.
+        Module::Initialize(executablePath);
+        v8::Isolate::CreateParams create_params;
+        create_params.array_buffer_allocator = v8::ArrayBuffer::Allocator::NewDefaultAllocator();
+        v8::Isolate* isolate = v8::Isolate::New(create_params);
+
+        // Use the isolate within a scope.
+        {
+            v8::Isolate::Scope isolate_scope{isolate};
+            v8::HandleScope isolate_handle_scope{isolate};
+            v8::Local<v8::Context> context = v8::Context::New(isolate);
+            v8::Context::Scope context_scope{context};
+
+            Napi::Env env = Napi::Attach(context);
+            Dispatch([](Napi::Env env) {
+                Module::Instance().AddPlatformToJavascript(env);
+            });

```

Comment:
Hmmm I was getting a memory access exception when trying to access the JSRuntime NativeObject. I thought it was specifically because I wasn't on the JS thread so I used Dispatch and it worked. Maybe it could be an order of operations thing?